broken pipeline bioformats2raw and raw2ometiff in 0.3.0-rc3?

Hello,

I am trying to rebuild the pyramid that is already an ome.tif by converting all through the `bioformats2raw` and `raw2ometiff` pipeline. 

And I am getting this error:
```
  Exception in thread "main" picocli.CommandLine$ExecutionException: Error while calling command (com.glencoesoftware.pyramid.PyramidFromDirectoryWriter@8646db9): java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
  	at picocli.CommandLine.executeUserObject(CommandLine.java:1792)
  	at picocli.CommandLine.access$900(CommandLine.java:145)
  	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
  	at picocli.CommandLine$RunLast.handle(CommandLine.java:2144)
  	at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
  	at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:1968)
  	at picocli.CommandLine.parseWithHandlers(CommandLine.java:2349)
  	at picocli.CommandLine.parseWithHandler(CommandLine.java:2284)
  Causedat picocli.CommandLine.call(CommandLine.java:2560)
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.main(PyramidFromDirectoryWriter.java:210)
  Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
  	at java.util.ArrayList.rangeCheck(ArrayList.java:659)
  	at java.util.ArrayList.get(ArrayList.java:435)
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.initialize(PyramidFromDirectoryWriter.java:726)
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.call(PyramidFromDirectoryWriter.java:225)
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.call(PyramidFromDirectoryWriter.java:97)
  	at picocli.CommandLine.executeUserObject(CommandLine.java:1783)
  rk dir... 9 more
```

Command line used for bf2raw:
`/bf2raw/bioformats2raw-0.3.0-rc3/bin/bioformats2raw --max_workers 20  --resolutions 5 --no-hcs $ome_tif "${stem}"`
and raw2ometiff:
`/raw2tif/raw2ometiff-0.3.0-rc3/bin/raw2ometiff --max_workers 20 "${stem}" [new.tif]`

PS: the output of bf2raw is a folder with `0` and `OME` which seems good to me.

Any idea what I did wrong?

Many thanks for your help!

Best,
Tong

Sorry I should have mentioned that - yeah. I tried without. Same error message.
```
  Exception in thread "main" picocli.CommandLine$ExecutionException: Error while calling command (com.glencoesoftware.pyramid.PyramidFromDirectoryWriter@8646db9): java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
  	at picocli.CommandLine.executeUserObject(CommandLine.java:1792)
  	at picocli.CommandLine.access$900(CommandLine.java:145)
  	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
  	at picocli.CommandLine$RunLast.handle(CommandLine.java:2144)
  	at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
  	at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:1968)
  	at picocli.CommandLine.parseWithHandlers(CommandLine.java:2349)
  	at picocli.CommandLine.parseWithHandler(CommandLine.java:2284)
  Causedat picocli.CommandLine.call(CommandLine.java:2560)
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.main(PyramidFromDirectoryWriter.java:210)
  Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
  	at java.util.ArrayList.rangeCheck(ArrayList.java:659)
  	at java.util.ArrayList.get(ArrayList.java:435)
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.initialize(PyramidFromDirectoryWriter.java:726)
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.call(PyramidFromDirectoryWriter.java:225)
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.call(PyramidFromDirectoryWriter.java:97)
  	at picocli.CommandLine.executeUserObject(CommandLine.java:1783)
  rk dir... 9 more
```
and yes. There was a warning for bioformats2raw
```
OpenJDK 64-Bit Server VM warning: You have loaded library /tmp/opencv_openpnp6105402081124591895/nu/pattern/opencv/linux/x86_64/libopencv_java342.so which might have disabled stack guard. The VM will try to fix t
he stack guard now.
It's highly recommended that you fix the library with 'execstack -c <libfile>', or link it with '-z noexecstack'.
```

Thank you for the additional info.  I can now reproduce this error message:

```
Exception in thread "main" picocli.CommandLine$ExecutionException: Error while calling command (com.glencoesoftware.pyramid.PyramidFromDirectoryWriter@4a87761d): java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at picocli.CommandLine.executeUserObject(CommandLine.java:1792)
	at picocli.CommandLine.access$900(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2144)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
	at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:1968)
	at picocli.CommandLine.parseWithHandlers(CommandLine.java:2349)
	at picocli.CommandLine.parseWithHandler(CommandLine.java:2284)
	at picocli.CommandLine.call(CommandLine.java:2560)
	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.main(PyramidFromDirectoryWriter.java:210)
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:659)
	at java.util.ArrayList.get(ArrayList.java:435)
	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.initialize(PyramidFromDirectoryWriter.java:726)
	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.call(PyramidFromDirectoryWriter.java:225)
	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.call(PyramidFromDirectoryWriter.java:97)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1783)
	... 9 more
```

when running ```raw2ometiff test test-output.ome.tiff``` after ```bioformats2raw test.ome.tiff test --resolutions 5```.  However, I can't reproduce the error when running the same ```raw2ometiff``` command after ```bioformats2raw test.ome.tiff test --resolutions 5 --no-hcs```.  Can you try running ```bioformats2raw``` with ```--no-hcs``` again with a non-existent output directory, verify that ```raw2ometiff``` shows the same error, and send the complete output of ```ls -Rhl``` on the Zarr directory written by ```bioformats2raw```?

The error occurs without the ```--no-hcs``` flag because the ```Well``` in the original OME-XML has an incorrect ```Row``` (and possibly ```Column```) index:

```
<Plate ID="Plate:0" Name="RV_GON_01A5" ColumnNamingConvention="letter" RowNamingConvention="number" Columns="4" Rows="1" ExternalIdentifier="6f8a58d4-5b1e-4ecc-93fd-824a8d2c3468">
    <Description>PerkinElmer Slide Holder H.45 (up to 4 Slides), Sample below Slide</Description>
    <Well ID="Well:0" Row="1" Column="1">
      <WellSample ID="WellSample:0:1" Index="1" PositionX="900.911" PositionXUnit="µm" PositionY="-15916.0885" PositionYUnit="µm">
        <ImageRef ID="Image:0"/>
      </WellSample>
    </Well>
  </Plate>
```

For reference, I created the test input file by pasting the OME-XML above into a file ```62.ome.xml```, and then running:

```
$ bfconvert "test&sizeC=5&pixelType=uint16&sizeX=35033&sizeY=30924.fake" test.ome.tiff -pyramid-resolutions 7 -tilex 512 -tiley 512 -compression LZW
$ tiffcomment -set 62.ome.xml test.ome.tiff
``` 

This warning isn't related, and is fine to ignore:

```
OpenJDK 64-Bit Server VM warning: You have loaded library /tmp/opencv_openpnp6105402081124591895/nu/pattern/opencv/linux/x86_64/libopencv_java342.so which might have disabled stack guard. The VM will try to fix t
he stack guard now.
It's highly recommended that you fix the library with 'execstack -c <libfile>', or link it with '-z noexecstack'.
```

So what I can tell is that the same file can be converted using `0.2.6` correctly (both bioformats2raw and raw2ometiff).
I've also tried bioformats2raw-0.2.6 with `file-type zarr` + raw2ometiff-0.3.0-rc3, and got this error:

```
  Exception in thread "main" picocli.CommandLine$ExecutionException: Error while calling command (com.glencoesoftware.pyramid.PyramidFromDirectoryWriter@8646db9): java.lang.RuntimeException: java.io.IOException: '.zgroup' expected but is not readable or missing in store.
  	at picocli.CommandLine.executeUserObject(CommandLine.java:1792)
  	at picocli.CommandLine.access$900(CommandLine.java:145)
  	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
  	at picocli.CommandLine$RunLast.handle(CommandLine.java:2144)
  	at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
  	at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:1968)
  	at picocli.CommandLine.parseWithHandlers(CommandLine.java:2349)
  Causedat picocli.CommandLine.parseWithHandler(CommandLine.java:2284)
  	at picocli.CommandLine.call(CommandLine.java:2560)
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.main(PyramidFromDirectoryWriter.java:210)
  Caused by: java.lang.RuntimeException: java.io.IOException: '.zgroup' expected but is not readable or missing in store.
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.call(PyramidFromDirectoryWriter.java:239)
  Causedat com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.call(PyramidFromDirectoryWriter.java:97)
  	at picocli.CommandLine.executeUserObject(CommandLine.java:1783)
  	... 9 more
  Caused by: java.io.IOException: '.zgroup' expected but is not readable or missing in store.
  	at com.bc.zarr.ZarrGroup.open(ZarrGroup.java:104)
  	at com.bc.zarr.ZarrGroup.open(ZarrGroup.java:95)
  	at com.bc.zarr.ZarrGroup.open(ZarrGroup.java:87)
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.createReader(PyramidFromDirectoryWriter.java:1150)
  	at com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.initialize(PyramidFromDirectoryWriter.java:576)
  rk dirat com.glencoesoftware.pyramid.PyramidFromDirectoryWriter.call(PyramidFromDirectoryWriter.java:225)
  /home/... 11 more
```
I assume this is expected though.

It should be me thanking you guys for this great tool. Just trying to help as more as I can.

As for the path, I can confirm that it's the correct path. I delete the directory with `rm -rf` for each test.

Just a question about debuging, which level should I use? The `ALL` gives a file of 33M and stopped by this error in `bioformats2raw`:
```
  Exception in thread "main" picocli.CommandLine$ExecutionException: Error while calling command (com.glencoesoftware.bioformats2raw.Converter@544a2ea6): java.lang.NullPointerException
  	at picocli.CommandLine.executeUserObject(CommandLine.java:1792)
  	at picocli.CommandLine.access$900(CommandLine.java:145)
  	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
  	at picocli.CommandLine$RunLast.handle(CommandLine.java:2144)
  	at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
  	at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:1968)
  	at picocli.CommandLine.parseWithHandlers(CommandLine.java:2349)
  	at picocli.CommandLine.parseWithHandler(CommandLine.java:2284)
  	at picocli.CommandLine.call(CommandLine.java:2560)
  	at com.glencoesoftware.bioformats2raw.Converter.main(Converter.java:1613)
  Caused by: java.lang.NullPointerException
  	at loci.common.RandomAccessInputStream.getFilePointer(RandomAccessInputStream.java:236)
  	at loci.formats.tiff.OnDemandLongArray.toArray(OnDemandLongArray.java:79)
  	at loci.formats.tiff.IFD.printIFD(IFD.java:980)
  	at loci.formats.tiff.TiffParser.getSamples(TiffParser.java:900)
  	at loci.formats.tiff.TiffParser.getSamples(TiffParser.java:869)
  	at loci.formats.in.OMETiffReader.openBytes(OMETiffReader.java:349)
  	at loci.formats.ReaderWrapper.openBytes(ReaderWrapper.java:348)
  	at loci.formats.ChannelSeparator.openBytes(ChannelSeparator.java:229)
  	at loci.formats.ChannelSeparator.openBytes(ChannelSeparator.java:161)
  	at com.glencoesoftware.bioformats2raw.Converter.getTile(Converter.java:863)
  	at com.glencoesoftware.bioformats2raw.Converter.processTile(Converter.java:959)
  	at com.glencoesoftware.bioformats2raw.Converter.lambda$saveResolutions$4(Converter.java:1109)
  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  	at java.lang.Thread.run(Thread.java:748)
  ```

--- a/build.gradle
+++ b/build.gradle
@@ -39,6 +39,7 @@ dependencies {
     // be careful here, this is a newer version of org.json:json than formats-gpl uses
     implementation 'org.json:json:20190722'
     testImplementation 'junit:junit:4.12'
+    testImplementation 'org.apache.commons:commons-lang3:3.12.0'
     testImplementation 'com.glencoesoftware:bioformats2raw:0.3.0-SNAPSHOT'
 }
 

--- a/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
+++ b/src/main/java/com/glencoesoftware/pyramid/PyramidFromDirectoryWriter.java
@@ -212,6 +212,8 @@ public class PyramidFromDirectoryWriter implements Callable<Void> {
 
   @Override
   public Void call() throws Exception {
+    // Resolve symlinks
+    inputDirectory = inputDirectory.toRealPath();
     if (printVersion) {
       printVersion();
       return null;

--- a/src/test/java/com/glencoesoftware/raw2ometiff/test/ConversionTest.java
+++ b/src/test/java/com/glencoesoftware/raw2ometiff/test/ConversionTest.java
@@ -241,14 +241,7 @@ public class ConversionTest {
         reader.isLittleEndian());
   }
 
-  /**
-   * Test defaults.
-   */
-  @Test
-  public void testDefaults() throws Exception {
-    input = fake();
-    assertBioFormats2Raw();
-    assertTool();
+  private void assertDefaults() throws Exception {
     ZarrArray series0 = ZarrGroup.open(output.resolve("0")).openArray("0");
     Assert.assertTrue(series0.getNested());
     // Also ensure we're using the latest .zarray metadata
@@ -266,6 +259,31 @@ public class ConversionTest {
 
   /**
    * Test defaults.
+   */
+  @Test
+  public void testDefaults() throws Exception {
+    input = fake();
+    assertBioFormats2Raw();
+    assertTool();
+    assertDefaults();
+  }
+
+  /**
+   * Test symlink as the root.
+   */
+  @Test
+  public void testSymlinkAsRoot() throws Exception {
+    input = fake();
+    assertBioFormats2Raw();
+    Path notASymlink = output.resolveSibling(output.getFileName() + ".old");
+    Files.move(output, notASymlink);
+    Files.createSymbolicLink(output, notASymlink);
+    assertTool();
+    assertDefaults();
+  }
+
+  /**
+   * Test series count check.
    */
   @Test
   public void testSeriesCountCheck() throws Exception {

--- a/src/test/java/com/glencoesoftware/raw2ometiff/test/ConversionTest.java
+++ b/src/test/java/com/glencoesoftware/raw2ometiff/test/ConversionTest.java
@@ -40,7 +40,9 @@ import loci.formats.tiff.TiffParser;
 import picocli.CommandLine;
 import picocli.CommandLine.ExecutionException;
 
+import org.apache.commons.lang3.SystemUtils;
 import org.junit.Assert;
+import org.junit.Assume;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
@@ -273,6 +275,7 @@ public class ConversionTest {
    */
   @Test
   public void testSymlinkAsRoot() throws Exception {
+    Assume.assumeTrue(SystemUtils.IS_OS_LINUX);
     input = fake();
     assertBioFormats2Raw();
     Path notASymlink = output.resolveSibling(output.getFileName() + ".old");

